### PR TITLE
Add deque structure

### PR DESCRIPTION
**Pull request by bose-sayan**
Hey everyone,

### Implemented Doubly-Ended Queue (Deque) Data Structure

I've implemented a Doubly-Ended Queue (Deque) data structure in this PR. A Deque allows elements to be added or removed from both ends, making it a versatile data structure for various use cases.

Here's a brief overview of what I've done:

- Created a `Deque` struct with methods for:
  - Adding elements to the front and rear.
  - Removing elements from the front and rear.
  - Checking if the Deque is empty.
  - Getting the size of the Deque.
  
- Implemented the necessary underlying data structure to support Deque operations efficiently.

### Added Test Cases for Correctness

Ensuring the correctness of our data structure is crucial. To do this, I've added a set of test cases that thoroughly evaluate the functionality of the Deque. These tests cover scenarios like:

- Adding and removing elements from both ends.
- Checking the size of the Deque.
- Testing edge cases and boundary conditions.
- Verifying that the Deque behaves as expected when empty.

### How to Review

To review the changes, please take a look at the code in the `deque.go` file to see the implementation and the `deque_test.go` file for the test cases.

Your feedback and suggestions are highly appreciated! If you spot any issues or have any recommendations for improvements, please don't hesitate to let me know. Together, we can make our codebase even stronger.

Thanks for your time and consideration!

Best regards,
Sayan 😊

### CODE REVIEW

File: structure/deque/deque.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
+	Package deque implements a Double Ended Queue data structure.
```
Comment (raklaptudirm) [inline structure/deque/deque.go:2]: Package comment will probably get converted to code block, remove the unnecessary indentation.

File: structure/deque/deque.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+	Author: [Sayan](https://github.com/bose-sayan)
+*/
+
+package deque
+
+import "fmt"
+
+type DoublyEndedQueue struct {
+	Deque []any
```
Comment (raklaptudirm) [inline structure/deque/deque.go:30]: This variable should be private and unaccessible by other packages. Also, instead of `any`, add a generic parameter to the type.

File: structure/deque/deque_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
+// Tests for the deque package.
+
+package deque
```
Comment (raklaptudirm) [inline structure/deque/deque_test.go:3]: Do black box testing in the package `deque_test`.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]
- Use `errors.New()` for constant errors. Also instead of calling the error creation function every time declare global variables and use those.
- Instead of separate empty errors for each function, a single `ErrEmptyDeQueue` should suffice.
- Export the methods of the `DoubleEndedQueue`.
- Go naming convention follows that getters should be named `Field` instead of `GetField`, so change the names for front, rear, and others.

File: structure/deque/deque_test.go [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+
+import (
+	"testing"
+)
+
+// Zero values
+var zeroInt int
```
Comment (raklaptudirm) [inline structure/deque/deque_test.go:10]: Unnecessary?

File: structure/deque/deque_test.go [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+func TestDeque(t *testing.T) {
+
+	// Test cases with ints as params
+	testCasesInt := []TestCaseData[int]{
+		{
+			description: "Test empty deque",
+			queries: []QueryStructure[int]{
+				{
+					queryType:      "IsEmpty",
+					parameter:      zeroInt,
```
Comment (raklaptudirm) [inline structure/deque/deque_test.go:34]: Also unnecessary, they are literally the zero or default values.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

File: structure/deque/deque_test.go (hunk 4 above)
Comment (bose-sayan) [reply structure/deque/deque_test.go:10]: Yeah, my bad. Sorry about that!

File: structure/deque/deque_test.go (hunk 3 above)
Comment (raklaptudirm) [reply structure/deque/deque_test.go:3]: The tests are still in the `deque` package? How is this resolved?

Comment (bose-sayan) [reply structure/deque/deque_test.go:3]: Hey, sorry about that. I don't get what is asked. 
Cause I was looking at the other test codes and wrote my one accordingly.
Could you clarify please? It would be really helpful

Comment (raklaptudirm) [reply structure/deque/deque_test.go:3]: You should write the tests under the package `deque_test` instead of `deque` cause you are testing the public API.

https://github.com/TheAlgorithms/Go/blob/679acf08f4b43385e2cb2f13ff0c8a93a6100027/structure/deque/deque_test.go#L3

Comment (bose-sayan) [reply structure/deque/deque_test.go:3]: Ah I see, got it now. Thanks for the clarification!

File: structure/deque/deque_test.go [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+	expectedResult interface{}
+	expectedError  error
+}
+
+type TestCaseData[T any] struct {
+	description string
+	queries     []QueryStructure[T]
+}
+
+var ErrEmptyDeQueue = deque.ErrEmptyDeQueue
```
Comment (raklaptudirm) [inline structure/deque/deque_test.go:36]: Why redeclare the variable?

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

File: structure/deque/deque_test.go (hunk 6 above)
Comment (bose-sayan) [reply structure/deque/deque_test.go]: Did the needed modification. Thanks!

**Review by raklaptudirm — DISMISSED** [review]

File: structure/deque/deque.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	getRear(): Gets the last item from queue.
+	isEmpty(): Checks whether Deque is empty or not.
+	getLength(): Gets the length of Deque.
+
+References:
+	Wikipedia : https://en.wikipedia.org/wiki/Double-ended_queue
+	Github: https://www.geeksforgeeks.org/deque-set-1-introduction-applications/
+
+Author: [Sayan](https://github.com/bose-sayan)
+*/
```
Comment (tjgurwara99) [inline structure/deque/deque.go:23]: This comment style would not be detected by any of the go based documentation tools. Please read the guidance on how to document go code [here](https://go.dev/blog/godoc) 🙏🏼

Comment (bose-sayan) [reply structure/deque/deque.go]: Hey! Thanks for the clarification. I refactored the comments accordingly, but am not sure if the style aligns with how go doc expects it to be. It would be really great if you could check it out once. Thank you again!

File: structure/deque/deque.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+//	DeQueueFront(): Deletes an item from front of Deque.
+//	DeQueueRear(): Deletes an item from rear of Deque.
+//	Front(): Gets the front item from queue.
+//	Rear(): Gets the last item from queue.
+//	IsEmpty(): Checks whether Deque is empty or not.
+//	Length(): Gets the length of Deque.
+// References:
+//	Wikipedia : https://en.wikipedia.org/wiki/Double-ended_queue
+//	Github: https://www.geeksforgeeks.org/deque-set-1-introduction-applications/
+// author [Sayan](https://github.com/bose-sayan)
```
Comment (tjgurwara99) [inline structure/deque/deque.go:14]: You don't need to provide a list of all operations here tbh... I'm approving it anyways since its just a preference thing.

**Review by tjgurwara99 — APPROVED** [review]

File: structure/deque/deque.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+// Package deque implements a Double Ended Queue data structure.
+package deque
+
+import (
+	"errors"
+)
+
+// Custom error for handling cases when some dequeuing operation is performed on an empty deque.
+var ErrEmptyDeQueue = errors.New("DoublyEnded queue is empty, so can't perform this operation")
```
Comment (tjgurwara99) [inline structure/deque/deque.go:24]:
```suggestion
// ErrEmptyDeQueue is a custom error for handling cases when some dequeuing operation is performed on an empty deque.
var ErrEmptyDeQueue = errors.New("DoublyEnded queue is empty, so can't perform this operation")
```

File: structure/deque/deque_test.go [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+// This file contains unit tests for the deque package.
+// The tests cover the following scenarios:
+// - Empty deque
+// - Deque with one element
+// - Deque with multiple elements
+// The tests are parameterized with int and string types.
+// Each test case is defined with a description and a list of queries to be executed on the deque.
+// The expected results and errors are also defined for each query.
+//
+// To run the tests, execute the TestDeque function.
```
Comment (tjgurwara99) [inline structure/deque/deque_test.go:12]: This is not necessary
```suggestion
```

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
I noticed something after I approved it previously.

File: structure/deque/deque.go (hunk 8 above)
Comment (bose-sayan) [reply structure/deque/deque.go:14]: Thanks for the suggestions! 👍

File: structure/deque/deque.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+// ErrEmptyDeQueue is a custom error for handling cases when some dequeuing operation is performed on an empty deque.
+var ErrEmptyDeQueue = errors.New("DoublyEnded queue is empty, so can't perform this operation")
+
+type DoublyEndedQueue[T any] struct {
+	deque []T
+}
+
+// NewDeque returns a new DoublyEndedQueue.
+func NewDeque[T any]() *DoublyEndedQueue[T] {
```
Comment (tjgurwara99) [inline structure/deque/deque.go:22]:
```suggestion
func New[T any]() *DoublyEndedQueue[T] {
```

`deque` is in the name of the package so no need to repeat it here.

File: structure/deque/deque.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	return &DoublyEndedQueue[T]{deque: make([]T, 0)}
+}
+
+// EnQueueFront adds an item at the front of Deque.
+func (dq *DoublyEndedQueue[T]) EnQueueFront(item T) {
```
Comment (tjgurwara99) [inline structure/deque/deque.go:27]:
```suggestion
func (dq *DoublyEndedQueue[T]) EnqueueFront(item T) {
```

Comment (tjgurwara99) [reply structure/deque/deque.go:27]: Same for other `EnQueue*` methods - `Enqueue` is one word

File: structure/deque/deque.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+	dq.deque = append([]T{item}, dq.deque...)
+}
+
+// EnQueueRear adds an item at the rear of Deque.
+func (dq *DoublyEndedQueue[T]) EnQueueRear(item T) {
+	dq.deque = append(dq.deque, item)
+}
+
+// DeQueueFront deletes an item from front of Deque and returns it.
+func (dq *DoublyEndedQueue[T]) DeQueueFront() (T, error) {
```
Comment (tjgurwara99) [inline structure/deque/deque.go:37]:
```suggestion
func (dq *DoublyEndedQueue[T]) DequeueFront() (T, error) {
```

Comment (tjgurwara99) [reply structure/deque/deque.go:37]: Same for other `DeQueue*` methods

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: structure/deque/deque.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+// Package deque implements a Double Ended Queue data structure.
+package deque
+
+import (
+	"errors"
+)
+
+// ErrEmptyDeQueue is a custom error for handling cases when some dequeuing operation is performed on an empty deque.
+var ErrEmptyDeQueue = errors.New("DoublyEnded queue is empty, so can't perform this operation")
```
Comment (tjgurwara99) [inline structure/deque/deque.go:15]:
```suggestion
// ErrEmptyDequeue is a custom error for handling cases when some dequeuing operation is performed on an empty deque.
var ErrEmptyDequeue = errors.New("DoublyEnded queue is empty, so can't perform this operation")
```
you missed these 😄

Comment (bose-sayan) [reply structure/deque/deque.go:15]: Sorry about that 😅

**Review by tjgurwara99 — APPROVED** [review]

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
You seem to have only changed the definition - the usage is still calling `ErrEmptyDeQueue` instead of `ErrEmptyDequeue`...

If you're using editors, there are features that allow you to rename in one go...

**Comment by bose-sayan** [comment]
@tjgurwara99 Sorry about that! I messed up in a hurry 🥲

File: structure/deque/deque_test.go [hunk 15]
```diff
@@ -0,0 +1,407 @@
+//
+// Package deque provides a double-ended queue data structure implementation.
```
Comment (tjgurwara99) [inline structure/deque/deque_test.go:2]: Do you need to repeat yourself again here? I think not 😄 
```suggestion
```

**Review by tjgurwara99 — APPROVED** [review]

File: structure/deque/deque_test.go (hunk 15 above)
Comment (bose-sayan) [reply structure/deque/deque_test.go:2]: Yeah true, lemme edit it quickly!

**Review by tjgurwara99 — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]
